aws/resource_aws_ecr_repository_policy_statement.go

```diff
@@ -0,0 +1,294 @@
+	return resourceAwsEcrRepositoryPolicyStatementCreateOrUpdate(d, meta)
+}
+
+func resourceAwsEcrRepositoryPolicyStatementDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ecrconn
+	sid := d.Get("sid").(string)
+	registryID := d.Get("registry_id").(string)
+	repository := d.Get("repository").(string)
+
+	awsMutexKV.Lock(repository)
+	defer awsMutexKV.Unlock(repository)
+
+	log.Printf("[DEBUG] Reading repository policy statement %s", d.Id())
+
+	_, policy, err := readOrBuildEcrPolicy(conn, registryID, repository)
+	if err != nil {
+		return err
+	}
+
+	index, found := findEcrPolicyStatement(policy, sid)
+	if !found {
+		d.SetId("")
+		return nil
+	}
+
+	policy.Statements = append(policy.Statements[:index], policy.Statements[index+1:]...)
+
+	_, err = resourceAwsEcrRepositoryPolicyStatementUpdateWith(d, meta, policy)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] repository policy statement %s deleted.", d.Id())
+
+	return nil
+}
+
+func resourceAwsEcrRepositoryPolicyStatementUpdateWith(d *schema.ResourceData, meta interface{}, policy *ecrPolicy) (*ecr.SetRepositoryPolicyOutput, error) {
+	conn := meta.(*AWSClient).ecrconn
+	registryID := d.Get("registry_id").(string)
+	repository := d.Get("repository").(string)
+
+	if len(policy.Statements) == 0 {
+		input := ecr.DeleteRepositoryPolicyInput{
+			RepositoryName: aws.String(repository),
+		}
+
+		if registryID != "" {
+			input.RegistryId = aws.String(registryID)
+		}
+
+		_, err := conn.DeleteRepositoryPolicy(&input)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+	}
+
+	policyText, err := json.Marshal(policy)
+	if err != nil {
+		return nil, err
+	}
+
+	input := ecr.SetRepositoryPolicyInput{
+		RepositoryName: aws.String(repository),
+		PolicyText:     aws.String(string(policyText)),
+	}
+
+	if registryID != "" {
+		input.RegistryId = aws.String(registryID)
+	}
+
+	out, err := conn.SetRepositoryPolicy(&input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func readOrBuildEcrPolicy(conn *ecr.ECR, registryID string, repositoryName string) (*ecr.GetRepositoryPolicyOutput, *ecrPolicy, error) {
+	input := &ecr.GetRepositoryPolicyInput{
+		RepositoryName: aws.String(repositoryName),
+	}
+
+	if registryID != "" {
+		input.RegistryId = aws.String(registryID)
+	}
+
+	out, err := conn.GetRepositoryPolicy(input)
+
+	if err != nil {
+		if ecrerr, ok := err.(awserr.Error); ok {
+			switch ecrerr.Code() {
+			case "RepositoryNotFoundException", "RepositoryPolicyNotFoundException":
+				// Do nothing
+			default:
+				return nil, nil, err
+			}
+		} else {
+			return nil, nil, err
+		}
+	}
+
+	policy := &ecrPolicy{}
+	if out != nil && out.PolicyText != nil {
+		if err := json.Unmarshal([]byte(*out.PolicyText), policy); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		policy.Version = "2008-10-17"
+	}
+
+	if policy.Statements == nil {
+		policy.Statements = make([]*ecrPolicyStatement, 0)
+	}
```

Comment:
@greenboxal Do you think you could take some time to create a data source called `aws_iam_policy_document` which would have a source_policy document, that would allow to merge your policy into the source one?

Your use case seems totally valid and it would be a great addition to have it! 👍 
I am just wondering that maybe we could make this part reusable so that we can handle this case in several situations. What do you think? :)

Reply:
@Ninir this sounds almost exactly like what I did with #2890